servers/src/grin/sync.rs

```diff
@@ -516,16 +519,31 @@ fn get_locator(
 ) -> Result<Vec<Hash>, Error> {
 	let mut this_height = 0;
 
-	let tip = chain.get_sync_head()?;
-	let heights = get_locator_heights(tip.height);
-	let mut new_heights: Vec<u64> = vec![];
+	let mut tip = chain.get_sync_head()?;
+	let header_head = chain.get_header_head()?;
 
 	// for security, clear history_locators[] in any case of header chain rollback,
 	// the easiest way is to check whether the sync head and the header head are identical.
-	if history_locators.len() > 0 && tip.hash() != chain.get_header_head()?.hash() {
+	if history_locators.len() > 0 && tip.hash() != header_head.hash() {
 		history_locators.clear();
 	}
 
+	// reset to header_head if sync_head is left behind
+	if header_head.height > tip.height {
+		debug!(
+			LOGGER,
+			"sync: get_locator reset to header_head. sync_head: {} at {}, header_head: {} at {}",
+			tip.hash(),
+			tip.height,
+			header_head.hash(),
+			header_head.height,
+		);
+		tip = header_head.clone();
```

Comment:
Let me put a PR up to illustrate what I mean.

Reply:
OK @antiochp  ,wait your PR.
but I'm not sure it's necessary here to update `sync_head`, since we will get the Headers downloaded and update `sync_head` when handle the received Headers.

Reply:
The point is we want to do this first.
We're starting a new header sync - we should put the sync_head in the correct state up front.
There's no point sending and receiving 500 headers from another peer, just to then update the stale sync_head to the header_head (that we already know).

Reply:
> There's no point sending and receiving 500 headers from another peer, just to then update the stale sync_head to the header_head (that we already know).

No, since we have use `header_head` for `get_locator()`, the new received Headers will all new for us, no duplicated Headers will be sent.

> Or temporarily overriding it when we know it will get updated next iteration.

Sure we know it will get updated next iteration, but this is not a temporarily overriding, it's for the purpose to keep `sync_head` ONLY updated when we received NEW Headers, and we can always count on this `sync_head` to convince us what's the latest Headers we received.